base/MAPL_ISO8601_Time.F90

```diff
@@ -0,0 +1,720 @@
+   end type ISO8601Date
+
+   interface ISO8601Date
+      module procedure :: construct_ISO8601Date
+   end interface ISO8601Date
+
+   type :: ISO8601Time
+      type(time_fields) :: fields
+      integer :: timezone
+   end type ISO8601Time
+
+   interface ISO8601Time
+      module procedure :: construct_ISO8601Time
+   end interface ISO8601Time
+
+   type :: ISO8601DateTime
+      type(ISO8601Date) :: date
+      type(ISO8601Time) :: time
+   contains
+      procedure, pass(self) :: year
+      procedure, pass(self) :: month
+      procedure, pass(self) :: day
+      procedure, pass(self) :: hour
+      procedure, pass(self) :: minute
+      procedure, pass(self) :: second
+      procedure, pass(self) :: millisecond
+      procedure, pass(self) :: timezone
+   end type ISO8601DateTime
+
+   interface ISO8601DateTime
+      module procedure :: construct_ISO8601DateTime
+   end interface ISO8601DateTime
+
+   type :: ISO8601Duration
+      type(date_fields) :: date
+      type(time_fields) :: time
+   end type ISO8601Duration
+
+   type :: ISO8601TimeInterval
+      type(ISO8601DateTime) :: start_datetime
+      type(ISO8601DateTime) :: end_datetime
+      integer :: repetitions
+   end type ISO8601TimeInterval
+
+contains
+
+
+! SIMPLE CONSTRUCTORS
+
+   pure function construct_date_fields(fields) result(df)
+      integer, dimension(DIM_DF), intent(in) :: fields
+      type(date_fields) :: df
+      df % year = fields(1)
+      df % month = fields(2)
+      df % day = fields(3)
+      df % valid = is_valid_date(fields)
+   end function construct_date_fields
+
+   pure function construct_time_fields(fields) result(tf)
+      integer, dimension(DIM_TF), intent(in) :: fields
+      type(time_fields) :: tf
+      tf % hour = fields(1)
+      tf % minute = fields(2)
+      tf % second = fields(3)
+      tf % millisecond = fields(4)
+      tf % timezone = fields(5)
+      tf % valid = is_valid_time(fields)
+   end function construct_time_fields
+
+! END SIMPLE CONSTRUCTORS
+
+
+! LOW-LEVEL UTILITY FUNCTIONS FOR STRING PARSERS
+
+   pure function make_char_array(s) result(a)
+      character(len=*), intent(in) :: s
+      character, dimension(len(s)) :: a
+      integer :: i
+      do i=1, len(s)
+         a(i) = s(i:i)
+      end do
+   end function make_char_array
+
+   elemental pure logical function is_digit(c)
+      character, intent(in) :: c
+      integer, parameter :: lb = iachar('0') - 1
+      integer, parameter :: ub = iachar('9') + 1
+      is_digit = iachar(c) > lb .and. iachar(c) < ub
+   end function is_digit
+
+   pure logical function are_digits(s)
+      character(len=*), intent(in) :: s
+      are_digits = all(is_digit(make_char_array(s)))
+   end function are_digits
+
+   pure logical function is_delimited_date(s)
+      character(len=*), intent(in) :: s
+      is_delimited_date = len(s)==10 .and. are_digits(s(1:4)) &
+         .and. s(5:5)==DD .and. are_digits(s(6:7)) .and. s(8:8)==DD &
+         .and. are_digits(s(9:10))
```

Comment:
Would be nice to avoid the magic numbers here.    Of course using named parameters will make the expression far worse, so may need a different approach.  Maybe something along the lines of:
```
is_delimited_date = (len(s)==???) .and.   are_digits(year_pos(s)) .and. is_separator(separator_pos(s)) .and. ...
```